server/src/grpc/remote_engine_service/mod.rs

```diff
@@ -97,6 +155,140 @@ impl<Q: QueryExecutor + 'static> RemoteEngineServiceImpl<Q> {
         Ok(ReceiverStream::new(rx))
     }
 
+    async fn deduped_stream_read_internal(
+        &self,
+        request: Request<ReadRequest>,
+    ) -> Result<ReceiverStream<Result<RecordBatch>>> {
+        let instant = Instant::now();
+        let ctx = self.handler_ctx();
+        let (tx, rx) = mpsc::channel(STREAM_QUERY_CHANNEL_LEN);
+
+        let request = request.into_inner();
+        let table_engine::remote::model::ReadRequest {
+            table,
+            read_request,
+        } = request.clone().try_into().box_err().context(ErrWithCause {
+            code: StatusCode::BadRequest,
+            msg: "fail to convert read request",
+        })?;
+
+        let request_key = RequestKey::new(
+            table.table,
+            read_request.predicate.clone(),
+            read_request.projected_schema.projection(),
+            read_request.order,
+        );
+
+        let mut _guard = RequestNotifierGuard::default();
+        match self
+            .request_notifiers
+            .as_ref()
+            .unwrap()
+            .insert_notifier(request_key.clone(), tx)
+        {
+            // The first request, need to handle it, and then notify the other requests.
+            RequestResult::First => {
+                // The _guard is used to remove key when future is cancelled.
+                _guard.set_some(|| {
```

Comment:
There is no need to use the pattern declare and update for the guard. A declaration is enough:
```rust
let guard = match ... {
    First => ExecutionGuard::new(..),
    .... => return ...,
}
```